middleware/ordered_group.go

```diff
@@ -0,0 +1,238 @@
+package middleware
+
+import "fmt"
+
+// RelativePosition provides specifying the relative position of a middleware
+// in an ordered group.
+type RelativePosition int
+
+// Relative position for middleware in steps.
+const (
+	After RelativePosition = iota
+	Before
+)
+
+type ider interface {
+	ID() string
+}
+
+// orderedIDs provides an ordered collection of items with relative ordering
+// by name.
+type orderedIDs struct {
+	order relativeOrder
+	items map[string]ider
+
+	unamedCounter int
+}
+
+func newOrderedIDs() *orderedIDs {
+	return &orderedIDs{
+		items: map[string]ider{},
+	}
+}
+
+func (g *orderedIDs) generateName() string {
+	g.unamedCounter++
+	return fmt.Sprintf("ordered group unnamed %d", g.unamedCounter)
+}
+
+// Add injects the item to the relative position of the item group. Returns an
+// error if the item already exists.
+//
+// If the value implements the ider interface, that id must be unique. If
+// the value does not implement the ider interface, a new unique id will be
+// used.
+func (g *orderedIDs) Add(m ider, pos RelativePosition) error {
+	if len(m.ID()) == 0 {
+		return fmt.Errorf("empty ID, ID must not be empty")
+	}
+
+	if err := g.order.Add(m.ID(), pos); err != nil {
+		return err
+	}
+
+	g.items[m.ID()] = m
+	return nil
+}
+
+// Insert injects the item relative to an existing item id.  Return error if
+// the original item does not exist, or the item being added already exists.
+//
+// If the value implements the ider interface, that id must be unique. If
+// the value does not implement the ider interface, a new unique id will be
+// used.
+func (g *orderedIDs) Insert(m ider, relativeTo string, pos RelativePosition) error {
+	if len(m.ID()) == 0 {
+		return fmt.Errorf("insert ID must not be empty")
+	}
+	if len(relativeTo) == 0 {
+		return fmt.Errorf("relative to ID must not be empty")
+	}
+
+	if err := g.order.Insert(m.ID(), relativeTo, pos); err != nil {
+		return err
+	}
+
+	g.items[m.ID()] = m
+	return nil
+}
+
+// Swap removes the item by id, replacing it with the new item. Returns error
+// if the original item doesn't exist.
+//
+// If the value implements the ider interface, that id must be unique. If
+// the value does not implement the ider interface, a new unique id will be
+// used.
+func (g *orderedIDs) Swap(id string, m ider) error {
```

Comment:
Should we return the original `ider` after a swap?

Reply:
sure can update to return the swapped out ider.